File: ast/ast.go
package ast

import (
	"github.com/ruistola/cooper/lexer"
)

type TypeExpr interface {
	typeExpr()
}

type Expr interface {
	expr()
}

type Stmt interface {
	stmt()
}

type NamedTypeExpr struct {
	TypeName string
}

func (t *NamedTypeExpr) typeExpr() {}

type ArrayTypeExpr struct {
	UnderlyingType TypeExpr
}

func (t *ArrayTypeExpr) typeExpr() {}

// PointerTypeExpr is the postfix `^` pointer type constructor, e.g. `T^`.
type PointerTypeExpr struct {
	UnderlyingType TypeExpr
}

func (t *PointerTypeExpr) typeExpr() {}

type FuncTypeExpr struct {
	ReturnType TypeExpr
	ParamTypes []TypeExpr
}

func (t *FuncTypeExpr) typeExpr() {}

type UnitTypeExpr struct{}

func (t *UnitTypeExpr) typeExpr() {}

// TypeApplicationExpr applies a type constructor to one or more argument types by
// juxtaposition, e.g. `Map string i32` or `Box i32`. The spine is parsed
// left-associatively; the constructor is typically a NamedTypeExpr and arity is
// checked later by the resolver. Postfix `[]`/`^` bind tighter than application,
// so compound arguments must be parenthesised (`(Map string i32)[]`).
type TypeApplicationExpr struct {
	Constructor TypeExpr
	Args        []TypeExpr
}

func (t *TypeApplicationExpr) typeExpr() {}

// TupleTypeExpr is an anonymous product type `(A, B, ...)` with at least two
// elements. A single-element parenthesised type collapses to that element and a
// zero-element one is the unit type, so those never produce a TupleTypeExpr.
type TupleTypeExpr struct {
	ElementTypes []TypeExpr
}

func (t *TupleTypeExpr) typeExpr() {}

type UnitExpr struct{}

func (e *UnitExpr) expr() {}

// TupleLiteralExpr is a tuple value `(a, b, ...)` with at least two elements. A
// single parenthesised expression is a GroupExpr and empty parens are UnitExpr,
// so those never produce a TupleLiteralExpr.
type TupleLiteralExpr struct {
	Elements []Expr
}

func (e *TupleLiteralExpr) expr() {}

type BoolLiteralExpr struct {
	Value bool
}

func (e *BoolLiteralExpr) expr() {}

type StringLiteralExpr struct {
	Value string
}

func (e *StringLiteralExpr) expr() {}

type IdentExpr struct {
	Value string
}

func (e *IdentExpr) expr() {}

type NumberLiteralExpr struct {
	Value string
}

func (e *NumberLiteralExpr) expr() {}

// NilLiteralExpr is the `nil` literal, the absent value of a pointer type.
type NilLiteralExpr struct{}

func (e *NilLiteralExpr) expr() {}

// AddressOfExpr is the prefix `&` operator, yielding a pointer to its operand.
type AddressOfExpr struct {
	Operand Expr
}

func (e *AddressOfExpr) expr() {}

// DerefExpr is the postfix `^` operator, yielding the value a pointer points to.
type DerefExpr struct {
	Operand Expr
}

func (e *DerefExpr) expr() {}

type UnaryExpr struct {
	Operator lexer.Token
	Rhs      Expr
}

func (e *UnaryExpr) expr() {}

type BinaryExpr struct {
	Lhs      Expr
	Operator lexer.Token
	Rhs      Expr
}

func (e *BinaryExpr) expr() {}

type BlockExpr struct {
	Statements []Stmt
	ResultExpr Expr
}

func (e *BlockExpr) expr() {}

type BlockStmt struct {
	Statements []Stmt
}

func (s *BlockStmt) stmt() {}

type ExpressionStmt struct {
	Expr              Expr
	ExplicitSemicolon bool
}

func (s *ExpressionStmt) stmt() {}

type GroupExpr struct {
	Expr Expr
}

func (e *GroupExpr) expr() {}

type VarDeclStmt struct {
	Var     TypedIdent
	InitVal Expr
}

func (s *VarDeclStmt) stmt() {}

type TypedIdent struct {
	Name string
	Type TypeExpr
}

type FuncDeclStmt struct {
	Receiver   *TypedIdent // nil for free functions, non-nil for methods
	Name       string
	TypeParams []string // type-parameter binders declared after the name (free functions and methods)
	Parameters []*TypedIdent
	ReturnType TypeExpr
	Body       *BlockStmt
}

func (s *FuncDeclStmt) stmt() {}

type FuncCallExpr struct {
	Func Expr
	Args []Expr
}

func (e *FuncCallExpr) expr() {}

type StructDeclStmt struct {
	Name       string
	TypeParams []string // type-parameter binders declared after the name, e.g. `struct Map K V`
	Members    []*TypedIdent
}

func (s *StructDeclStmt) stmt() {}

// OneofDeclStmt declares a sum type (tagged union), e.g.
//
//	oneof Result T E {
//	  Ok(T),
//	  Err(E),
//	}
//
// Type-parameter binders follow the name by juxtaposition (as with structs). Each
// variant carries an ordered, comma-delimited list of positional payload slots;
// an empty list (`None`) is a payload-free variant.
type OneofDeclStmt struct {
	Name       string
	TypeParams []string
	Variants   []*VariantDef
}

func (s *OneofDeclStmt) stmt() {}

// VariantDef is a single variant of a sum type: a name plus an ordered list of
// positional payload slot types (empty for a payload-free variant).
type VariantDef struct {
	Name    string
	Payload []TypeExpr
}

type StructLiteralExpr struct {
	Struct  Expr
	Members []*MemberAssignExpr
}

func (e *StructLiteralExpr) expr() {}

type StructMemberExpr struct {
	Struct Expr
	Member *IdentExpr
}

func (e *StructMemberExpr) expr() {}

type ArrayIndexExpr struct {
	Array Expr
	Index Expr
}

func (e *ArrayIndexExpr) expr() {}

type IfExpr struct {
	Cond Expr
	Then Expr
	Else Expr
}

func (e *IfExpr) expr() {}

type IfStmt struct {
	Cond Expr
	Then Stmt
	Else Stmt
}

func (s *IfStmt) stmt() {}

type ForStmt struct {
	Init Stmt
	Cond Expr
	Iter *ExpressionStmt
	Body *BlockStmt
}

func (s *ForStmt) stmt() {}

type AssignExpr struct {
	Assigne       Expr
	Operator      lexer.Token
	AssignedValue Expr
}

func (e *AssignExpr) expr() {}

type MemberAssignExpr struct {
	Name  string
	Value Expr
}

func (e *MemberAssignExpr) expr() {}

type VarDeclAssignExpr struct {
	Name          string
	AssignedValue Expr
}

func (e *VarDeclAssignExpr) expr() {}

// TupleDeclAssignExpr destructures a tuple-typed initializer, declaring each name
// and binding it to the corresponding element positionally. It backs both the
// walrus form `(a, b) := rhs` (Type nil; element types inferred from rhs) and the
// `let (a, b): (A, B) = rhs` form (Type set; element types taken from and checked
// against the annotation). Patterns are untyped: any type annotation lives on the
// enclosing `let`, never inside the pattern.
type TupleDeclAssignExpr struct {
	Names         []string
	Type          TypeExpr
	AssignedValue Expr
}

func (e *TupleDeclAssignExpr) expr() {}

type ReturnStmt struct {
	Expr Expr
}

func (s *ReturnStmt) stmt() {}

type UseDeclStmt struct {
	UseSpecs []*UseSpecExpr
}

func (s *UseDeclStmt) stmt() {}

type UseSpecExpr struct {
	Alias string   // optional local alias; empty when absent
	Path  []string // period-separated module path segments (at least one)
}

func (e *UseSpecExpr) expr() {}

File: rust/src/ast.rs
use crate::lexer::TokenKind;

/// Type expressions. Mirrors the Go `ast.TypeExpr` interface, but as a closed
/// `enum` so every `match` over it is checked for exhaustiveness at compile time.
#[derive(Debug, Clone, PartialEq)]
pub enum TypeExpr {
    Named(String),
    Array(Box<TypeExpr>),
    /// Postfix `^` pointer constructor, e.g. `T^`.
    Pointer(Box<TypeExpr>),
    Func {
        return_type: Box<TypeExpr>,
        param_types: Vec<TypeExpr>,
    },
    Unit,
    /// Type application by juxtaposition, e.g. `Map string i32`.
    Application {
        constructor: Box<TypeExpr>,
        args: Vec<TypeExpr>,
    },
    /// Anonymous product type `(A, B, ...)` with at least two elements.
    Tuple(Vec<TypeExpr>),
}

/// A name paired with its declared type (struct member, parameter, receiver).
#[derive(Debug, Clone, PartialEq)]
pub struct TypedIdent {
    pub name: String,
    pub ty: TypeExpr,
}

/// A single variant of a sum type: a name plus positional payload slot types.
#[derive(Debug, Clone, PartialEq)]
pub struct VariantDef {
    pub name: String,
    pub payload: Vec<TypeExpr>,
}

/// A `name: value` initializer inside a struct literal.
#[derive(Debug, Clone, PartialEq)]
pub struct MemberAssign {
    pub name: String,
    pub value: Expr,
}

/// A value block: statements followed by a trailing result expression (unit when
/// the block produces no value).
#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub statements: Vec<Stmt>,
    pub result: Box<Expr>,
}

/// A match-arm pattern. Flat in this iteration: a type-qualified variant pattern
/// (optionally binding positional payload slots) or the wildcard `_`.
#[derive(Debug, Clone, PartialEq)]
pub enum Pattern {
    Variant {
        type_name: String,
        variant: String,
        binders: Vec<String>,
    },
    Wildcard,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchExprArm {
    pub pattern: Pattern,
    pub body: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchStmtArm {
    pub pattern: Pattern,
    pub body: Stmt,
}

/// Expressions.
#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Unit,
    Tuple(Vec<Expr>),
    Bool(bool),
    Str(String),
    Ident(String),
    Number(String),
    Nil,
    /// Prefix `&`.
    AddressOf(Box<Expr>),
    /// Postfix `^`.
    Deref(Box<Expr>),
    Unary {
        op: TokenKind,
        rhs: Box<Expr>,
    },
    Binary {
        lhs: Box<Expr>,
        op: TokenKind,
        rhs: Box<Expr>,
    },
    Block(Block),
    Group(Box<Expr>),
    FuncCall {
        func: Box<Expr>,
        args: Vec<Expr>,
    },
    StructLiteral {
        target: Box<Expr>,
        members: Vec<MemberAssign>,
    },
    StructMember {
        target: Box<Expr>,
        member: String,
    },
    ArrayIndex {
        array: Box<Expr>,
        index: Box<Expr>,
    },
    If {
        cond: Box<Expr>,
        then: Box<Expr>,
        els: Box<Expr>,
    },
    Match {
        scrutinee: Box<Expr>,
        arms: Vec<MatchExprArm>,
    },
    Assign {
        target: Box<Expr>,
        op: TokenKind,
        value: Box<Expr>,
    },
    VarDeclAssign {
        name: String,
        value: Box<Expr>,
    },
    TupleDeclAssign {
        names: Vec<String>,
        ty: Option<TypeExpr>,
        value: Box<Expr>,
    },
}

/// A free function or method declaration.
#[derive(Debug, Clone, PartialEq)]
pub struct FuncDecl {
    /// `None` for free functions, `Some` for methods.
    pub receiver: Option<TypedIdent>,
    pub name: String,
    pub type_params: Vec<String>,
    pub params: Vec<TypedIdent>,
    pub return_type: Option<TypeExpr>,
    pub body: Vec<Stmt>,
}

/// A `use` specifier: an optional local alias plus a dotted module path.
#[derive(Debug, Clone, PartialEq)]
pub struct UseSpec {
    pub alias: Option<String>,
    pub path: Vec<String>,
}

/// Statements.
#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    Block(Vec<Stmt>),
    Expression {
        expr: Expr,
        explicit_semicolon: bool,
    },
    VarDecl {
        name: String,
        ty: Option<TypeExpr>,
        init: Option<Expr>,
    },
    FuncDecl(FuncDecl),
    StructDecl {
        name: String,
        type_params: Vec<String>,
        members: Vec<TypedIdent>,
    },
    OneofDecl {
        name: String,
        type_params: Vec<String>,
        variants: Vec<VariantDef>,
    },
    If {
        cond: Expr,
        then: Box<Stmt>,
        els: Option<Box<Stmt>>,
    },
    For {
        init: Box<Stmt>,
        cond: Expr,
        iter: Expr,
        body: Vec<Stmt>,
    },
    Match {
        scrutinee: Expr,
        arms: Vec<MatchStmtArm>,
    },
    Return(Option<Expr>),
    Use(Vec<UseSpec>),
}
